drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode.rb

```diff
@@ -0,0 +1,194 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HudSpmReport::Fy2023
+  class Episode < GrdaWarehouseBase
+    self.table_name = 'hud_report_spm_episodes'
+    belongs_to :client, class_name: 'GrdaWarehouse::Hud::Client'
+
+    has_many :enrollment_links
+    has_many :enrollments, through: :enrollment_links
+    has_many :bed_nights
+
+    has_many :hud_reports_universe_members, inverse_of: :universe_membership, class_name: 'HudReports::UniverseMember', foreign_key: :universe_membership_id
+
+    attr_accessor :report # FIXME?
+
+    # The associations seem to make imports run one at a time, so, they are passed separately in parallel arrays
+    def self.save_episodes!(episodes, bed_nights, enrollment_links)
+      # Import the episodes
+      results = import!(episodes)
+      # Attach the associations to their episode
+      results.ids.each_with_index do |id, index|
+        bn_for_episode = bed_nights[index]
+        bed_nights[index] = bn_for_episode.map do |bn|
+          bn.episode_id = id
+          bn
+        end
+        el_for_episode = enrollment_links[index]
+        enrollment_links[index] = el_for_episode.map do |el|
+          el.episode_id = id
+          el
+        end
+      end
+      # Import the associations
+      BedNight.import!(bed_nights.flatten)
+      EnrollmentLink.import!(enrollment_links.flatten)
+    end
+
+    def compute_episode(enrollments, included_project_types:, excluded_project_types:, include_self_reported:)
+      raise 'Client undefined' unless client.present?
+
+      calculated_bed_nights = candidate_bed_nights(enrollments, included_project_types, include_self_reported)
+      calculated_excluded_dates = excluded_bed_nights(enrollments, excluded_project_types)
+      calculated_bed_nights.reject! { |_, _, date| date.in?(calculated_excluded_dates) }
+
+      filtered_bed_nights = filter_episode(calculated_bed_nights)
+      return unless filtered_bed_nights.present?
+
+      bed_nights_array = []
+      enrollment_links_array = []
+      filtered_bed_nights.each do |enrollment, service_id, date|
+        bed_nights_array << BedNight.new(
+          client_id: client.id,
+          enrollment_id: enrollment.id,
+          service_id: service_id,
+          date: date,
+        )
+      end
+
+      enrollment_ids = filtered_bed_nights.map { |enrollment, _, _| enrollment.id }.uniq
+      enrollment_ids.each do |enrollment_id|
+        enrollment_links_array << EnrollmentLink.new(
+          enrollment_id: enrollment_id,
+        )
+      end
+
+      dates = filtered_bed_nights.map(&:last)
+      first_date = dates.min
+      assign_attributes(
+        first_date: first_date,
+        last_date: dates.max,
```

Comment:
possible micro optimization, is filtered bed_nights be ordered already?  If it is then `dates.first` and `dates.last` would get you the min and max.  If not, could we just sort the list on line 70?  I'm unclear if a single sort is faster than a min and max, or there's also `first_date, last_date = dates.minmax`